// muse-js/dist/lib/muse-parse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("rxjs/add/operator/concatMap");
require("rxjs/add/operator/scan");
require("rxjs/add/operator/filter");
require("rxjs/add/operator/map");
function parseControl(controlData) {
    return controlData
        .concatMap(function (data) { return data.split(''); })
        .scan(function (acc, value) {
        if (acc.indexOf('}') >= 0) {
            return value;
        }
        else {
            return acc + value;
        }
    }, '')
        .filter(function (value) { return value.indexOf('}') >= 0; })
        .map(function (value) { return JSON.parse(value); });
}
exports.parseControl = parseControl;
function decodeUnsigned12BitData(samples) {
    var samples12Bit = [];
    for (var i = 0; i < samples.length; i++) {
        if (i % 3 === 0) {
            samples12Bit.push(samples[i] << 4 | samples[i + 1] >> 4);
        }
        else {
            samples12Bit.push((samples[i] & 0xf) << 8 | samples[i + 1]);
            i++;
        }
    }
    return samples12Bit;
}
exports.decodeUnsigned12BitData = decodeUnsigned12BitData;
function decodeEEGSamples(samples) {
    return decodeUnsigned12BitData(samples)
        .map(function (n) { return 0.48828125 * (n - 0x800); });
}
exports.decodeEEGSamples = decodeEEGSamples;
function parseTelemetry(data) {
    return {
        sequenceId: data.getUint16(0),
        batteryLevel: data.getUint16(2) / 512.,
        fuelGaugeVoltage: data.getUint16(4) * 2.2,
        // Next 2 bytes are probably ADC millivolt level, not sure
        temperature: data.getUint16(8),
    };
}
exports.parseTelemetry = parseTelemetry;
function parseAccelerometer(data) {
    function sample(startIndex) {
        return {
            x: data.getInt16(startIndex),
            y: data.getInt16(startIndex + 2),
            z: data.getInt16(startIndex + 4),
        };
    }
    return {
        sequenceId: data.getUint16(0),
        samples: [sample(2), sample(8), sample(14)]
    };
}
exports.parseAccelerometer = parseAccelerometer;
function parseGyroscope(data) {
    return parseAccelerometer(data);
}
exports.parseGyroscope = parseGyroscope;
